libclc/opencl/lib/amdgcn/mem_fence/fence.cl

```diff
@@ -8,11 +8,38 @@
 
 #include <clc/opencl/clc.h>
 
+void __clc_amdgcn_s_waitcnt(unsigned flags);
+
+// s_waitcnt takes 16bit argument with a combined number of maximum allowed
+// pending operations:
+// [12:8] LGKM -- LDS, GDS, Konstant (SMRD), Messages
+// [7] -- undefined
+// [6:4] -- exports, GDS, and mem write
+// [3:0] -- vector memory operations
+
+// Newer clang supports __builtin_amdgcn_s_waitcnt
+#if __clang_major__ >= 5
+#define __waitcnt(x) __builtin_amdgcn_s_waitcnt(x)
+#else
+#define __waitcnt(x) __clc_amdgcn_s_waitcnt(x)
+_CLC_DEF void __clc_amdgcn_s_waitcnt(unsigned) __asm("llvm.amdgcn.s.waitcnt");
+#endif
+
+_CLC_DEF _CLC_OVERLOAD void mem_fence(cl_mem_fence_flags flags) {
+  if (flags & CLK_GLOBAL_MEM_FENCE) {
```

Comment:
should we use similar implementation as in https://github.com/intel/llvm/blob/601222a5f7a226ba7e5d6ecd34f2fb4070a69b78/libclc/libspirv/lib/amdgcn-amdhsa/synchronization/barrier.cl#L51

Reply:
I'm not sure to be honest. If upstream `amdgcn` are okay with this implementation for OpenCL then I say we should just leave it as-is.